### ISSUE TITLE
Q: Does r_bg() return a failed process if R fails to start?

### ISSUE THREAD

**Issue by HenrikBengtsson** [r-lib/callr#214]
Hi. Assume that `./.Rprofile` produces an error during the R startup process, then `callr::r()` fails, e.g.

```r
library(callr)

## Run this example in the temporary folder
opwd <- setwd(tempdir())

## An .Rprofile that produces an error
cat("stop('boom')", file=".Rprofile")

v <- r(function() 42)
#> Error in get_result(output = out, options) : 
#>  callr subprocess failed: could not start R, exited with
#>  non-zero status, has crashed or was killed
```

Similarly with `r_bg()`:

```r
p <- r_bg(function() 42)
p$get_result()
#> Error in get_result(output = out, options) : 
#>  callr subprocess failed: could not start R, exited with
#>  non-zero status, has crashed or was killed
```

Now, when I inspect `p` (an `r_process` object), I can see that it failed:

```sh
p$is_alive()
#> [1] FALSE

p$get_exit_status()
#> [1] 1
```

My question is, does it fail already before `r_bg()` returns, or can it happen later?  I'm trying to decide if:

```r
p <- r_bg(function() 42)
if (p$get_exit_status() != 0) stop("Failed to launch R process")
```

is sufficient, or it could be that it fails afterwards and I have to wait for a while, e.g.

```r
p <- r_bg(function() 42)
Sys.sleep(10) ## wait long enough
if (p$get_exit_status() != 0) stop("Failed to launch R process")
```

Background: https://github.com/HenrikBengtsson/future/discussions/588#discussioncomment-2069582

**Comment by gaborcsardi** [comment]
If the R process could be started (i.e.`fork()` was successful) then it signals errors asynchronously:
```r
pr <- tempfile()
cat("stop()\n", file = pr)
Sys.setenv(R_PROFILE_USER = pr)
{ p <- r_bg(function() 42, user_profile=TRUE); p$get_exit_status() }
```

```r
#> NULL
```

(`$get_exit_status()` is `NULL` if the process is still running, so you cannot assume that it is an integer.)

Sleeping for a while is an anti-pattern, it is slow and still fragile. The good solution is to communicate back from the R process in some way, ideally in a way that the main process can poll. E.g. sending a message to the poll connection like `callr::r_session` does it, or printing some keyword to the standard output, and waiting for that in the main process.

The latter goes like this:

```r
start <- function(fun, ...) {
  p <- r_bg(fun, stderr = "2>&1", ...)

  grace <- as.difftime(5, units = "secs")
  limit <- Sys.time() + grace
  out <- ""

  while ((now <- Sys.time()) < limit && (incomplete <- p$is_incomplete_output())) {
    left <- as.double(limit - now, units = "secs")
    p$poll_io(as.integer(left * 1000))
    out <- paste0(out, p$read_output())
    if (grepl("FIDDLESTICKS", out))  break
  }

  # At this point we either got the keyword, or the process is done, or the time limit was up
  # If we didn't get the keyword, that's an error
  if (now >= limit) stop("Failed to start in ", format(grace))
  if (!incomplete) stop("Failed to start: ", out, p$read_all_output())

  p
}
```

The advantage of the polling is that 
* it is as fast as possible, no unnecessarily sleeping
* efficiently implemented by OS waiting on a fd or I/O Completion Ports
* you can wait for multiple processes at the same time if you want
* you can show a progress bar / spinner, if you want, if you poll in small time intervals

Some things to try:
```r
p <- start(function() { cat("FIDDLESTICKS\n"); Sys.sleep(10); 42 })
p$wait()
p$get_result()

start(function() { Sys.sleep(6) })
start(function() { stop("ooops") })
```

**Comment by HenrikBengtsson** [comment]
Thank you.

> Sleeping for a while is an anti-pattern, it is slow and still fragile.

Agree, and sorry for not being clear on that - I just used it to illustrate my question.

I like your `start()` approach. Part of me was hoping that `r_bg()` did some validation like that internally (=> before returning, `r_bg()` would know R had failed).  At the same time, I didn't think so because of the extra overhead introduced.

PS. What I'm after is for `f <- future(...)` with the **callr** backend to fail immediately, instead of as now, when `value(f)` is called. The closer to the source one can produce an error, the easier it is to troubleshoot.  I'll see if I use your polling technique for this, or just postpone the error, which should rather rare. But, this is not critical - failed R processes can happen at any time.
